plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs

```diff
@@ -235,112 +252,85 @@ machineInstance params =
+nextState oldState input = maybe (throwing _MSInvalidTransition (oldState, input)) pure (step oldState input)
+
+-- | Advance a running multisig contract. This applies the transition function
+--   'SM.transition' to the current contract state and uses the result to unlock
+--   the funds currently in the contract, and lock them again with a data script
+--   containing the new state.
+--
+runStepC :: AsMultiSigError e => Params -> Input -> Contract MultiSigSchema e ()
+runStepC params input = do
+    (oldState, value) <- currentState params
+    let addr = Scripts.scriptAddress $ scriptInstance params
+    utxos <- utxoAt addr
+    newState <- nextState oldState input
+    let ins = Typed.collectFromScript utxos (scriptInstance params) input
+        outs = Typed.makeScriptPayment (scriptInstance params) value newState
+    void 
+        $ withContractError (review _MSContractError)
+        $ writeTxSuccess 
+        $ ins <> outs
 
 -- | Lock some funds in a multisig contract.
-lock
-    :: (WalletAPI m, WalletDiagnostics m)
-    => Params
-    -- ^ Signatories and required signatures
-    -> Value
-    -- ^ The funds we want to lock
-    -> m State
-    -- ^ The initial state of the contract
-lock prms vl = do
-    (tx, state) <- SM.mkInitialise (machineInstance prms) (Holding vl) vl
-
-    void $ WAPITyped.signTxAndSubmit tx
-
-    pure state
-
--- | Propose a payment from funds that are locked up in a state-machine based
---   multisig contract.
-proposePayment
-    :: (WalletAPI m, WalletDiagnostics m)
-    => Params
-    -> State
-    -> Payment
-    -> m State
-proposePayment prms st = runStep prms st . ProposePayment
-
--- | Cancel a proposed payment
-cancelPayment
-    :: (WalletAPI m, WalletDiagnostics m)
-    => Params
-    -> State
-    -> m State
-cancelPayment prms st = runStep prms st Cancel
-
--- | Add a signature to a proposed payment
-addSignature
-    :: (WalletAPI m, WalletDiagnostics m)
-    => Params
-    -> State
-    -> m State
-addSignature prms st = WAPI.ownPubKey >>= runStep prms st . AddSignature
+lock :: AsMultiSigError e => Params -> Contract MultiSigSchema e ()
+lock params = do
+    value <- endpoint @"lock"
+    let tx = Typed.makeScriptPayment (scriptInstance params) value (Holding value)
+    void
+        $ withContractError (review _MSContractError)
+        $ writeTxSuccess tx
 
 -- | Make a payment after enough signatures have been collected.
-makePayment
-    :: (WalletAPI m, WalletDiagnostics m)
-    => Params
-    -> State
-    -> m State
-makePayment prms currentState = do
+makePaymentC :: AsMultiSigError e => Params -> Contract MultiSigSchema e ()
+makePaymentC params = do
+    () <- endpoint @"pay"
+    (state, _) <- currentState params
     -- we can't use 'runStep' because the outputs of the transaction are
     -- different from the other transitions: We need two outputs, a public
     -- key output with the payment, and the script output with the remaining
     -- funds.
```

Comment:
Could use `mkStep` or `mkHalt`?

Reply:
`mkStep` and `mkHalt` both have a `WalletAPI` constraint

Reply:
Yeah, they just need to be changed to take the `AddressMap` as an argument, which is how I've been gradually removing the `WalletAPI` constraints from everything ;)